File: services/api/app/schemas/text.py
from typing import Any

from pydantic import BaseModel, Field

MAX_TITLE_LENGTH = 300


class TextAnalysisCreate(BaseModel):
    text: str = Field(min_length=1)
    title: str | None = Field(default=None, max_length=MAX_TITLE_LENGTH)


class LanguageResponse(BaseModel):
    language: str | None
    confidence: float | None
    candidates: list[dict[str, Any]]
    engine: str
    reason: str | None


class TextAnalysisResponse(BaseModel):
    original_excerpt: str
    normalized_excerpt: str
    excerpt_chars: int
    truncated: bool
    original_sha256: str | None
    normalized_sha256: str
    normalization: dict[str, Any]
    language: LanguageResponse | None
    statistics: dict[str, Any]
    structure: dict[str, Any]
    limitations: list[str]

File: packages/shared-types/src/text.ts
/** Mirrors `app/schemas/text.py`. */

export interface TextAnalysisCreate {
  text: string;
  title?: string | null;
}

export interface LanguageGuess {
  language: string | null;
  /** Detector probability, 0–1. Not calibrated; a signal, not a fact. */
  confidence: number | null;
  candidates: Array<{ lang: string; prob: number }>;
  engine: string;
  reason: string | null;
}

export interface TextAnalysisResponse {
  original_excerpt: string;
  normalized_excerpt: string;
  excerpt_chars: number;
  truncated: boolean;
  original_sha256: string | null;
  normalized_sha256: string;
  normalization: Record<string, unknown>;
  language: LanguageGuess | null;
  statistics: Record<string, unknown>;
  structure: Record<string, unknown>;
  limitations: string[];
}

/** Input constraints enforced by the API (client checks are a courtesy only). */
export const TEXT_INPUT = { maxChars: 200_000 } as const;
